File: java/lesson 3/SubarraySum.java
import java.util.HashMap;
import java.util.Map;

// https://leetcode.com/problems/subarray-sum-equals-k/
class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> sums = new HashMap<>();
        sums.put(0, 1);
        
        int result = 0, sum = 0;
        for (int num : nums) {            
            sum += num;
            result += sums.getOrDefault(sum - k, 0);
            sums.put(sum, sums.getOrDefault(sum, 0) + 1);
        }
        
        return result;
    }
}

File: python/lesson 3/subarray_sum.py
# https://leetcode.com/problems/subarray-sum-equals-k/

from typing import List


class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        sums = { 0: 1 }
        result = sum = 0
        for num in nums:
            sum += num
            result += sums.get(sum - k, 0)
            sums[sum] = sums.get(sum, 0) + 1
        return result
